### PR TITLE
OnPolicyDriver

### PR DESCRIPTION
**Pull request by emailweixu**
OnPolicyDriver runs the eviroment with the algorithm for on-policy training. It's written
in such way that allow using tf.function for run(). So it's much faster than TrainingPolicy.

### CODE REVIEW

File: alf/drivers/on_policy_driver.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+            training_info = training_info._replace(
+                action_distribution=action_distribution)
+
+        if self._final_step_mode == OnPolicyDriver.FINAL_STEP_SKIP:
+            next_time_step, policy_step, _ = self._step(
+                time_step, policy_state)
+            next_state = policy_step.state
+        else:
+            policy_step = self._algorithm.train_step(time_step, policy_state)
+            next_time_step = time_step
```
Comment (witwolf) [inline alf/drivers/on_policy_driver.py]: why not perform it with new GradientTape and pass it to next _iter

File: alf/drivers/on_policy_driver.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+                self._algorithm.action_distribution_spec,
+                training_info.action_distribution)
+
+            training_info = training_info._replace(
+                action_distribution=action_distribution)
+
+        if self._final_step_mode == OnPolicyDriver.FINAL_STEP_SKIP:
+            next_time_step, policy_step, _ = self._step(
+                time_step, policy_state)
+            next_state = policy_step.state
```
Comment (witwolf) [inline alf/drivers/on_policy_driver.py]: why not perform final step with new GradientType and pass the new tape to next _iter and final_step_mode can be removed ?

Comment (witwolf) [reply alf/drivers/on_policy_driver.py]: loop_vars only support numpy array, `Tensor`, and `TensorArray` objects

Comment (emailweixu) [reply alf/drivers/on_policy_driver.py]: You said it. GradientType cannot be loop_vars

File: alf/algorithms/actor_critic_algorithm.py [hunk 3]
```diff
@@ -152,6 +151,8 @@ def calc_loss(self, training_info, final_time_step, final_policy_step):
 
             training_info = training_info._replace(
                 reward=training_info.reward + training_info.info.icm_reward)
+            final_time_step = final_time_step._replace(
+                reward=final_policy_step.info.icm_reward)
```
Comment (witwolf) [inline alf/algorithms/actor_critic_algorithm.py]: why is not 

     final_time_step = final_time_step._replace( 
              reward= final_time_step.reward+final_policy_step.info.icm_reward)

**Pull request merged**